manual/manual/tutorials/polymorphism.etex

```diff
@@ -0,0 +1,387 @@
+change manually the value stored by "store":
+\begin{caml_example}
+store := Some (fun x -> x);;
+\end{caml_example}
+Indeed, looking at the type of store, we see that the weak type "'_a" has
+been replaced by the type "int"
+\begin{caml_example}
+store;;
+\end{caml_example}
+Therefore, after placing an "int" in "store", we cannot use it to store any
+value other than an "int".
+
+More generally, weak types protect the program from undue mutation of
+values with a polymorphic type. Moreover, weak type cannot appear as the type
+of toplevel values: the type of values must be known at compilation time.
+Therefore, weak type must be determined by the end of the current toplevel
+module.
+
+\subsection{The value restriction}\label{ss:valuerestriction}
+
+Identifying the exact context in which polymorphic types should be
+replaced by weak types in a modular way is a difficult question. A first
+natural idea would be to state that any mutable object like reference, arrays
+or a record mutable field should never have a polymorphic type. Unfortunately,
+this rule is not sufficient, as illustrated by the following function that
+hides an uses an internal reference to implement a delayed identity function
+\begin{caml_example}
+let make_fake_id () =
+  let store = ref None in
+  fun x -> swap store x ;;
+let fake_id = make_fake_id();;
+\end{caml_example}
+It would be unsound to apply this fake_id function to values with different
+kinds. The function "fake_id" is therefore rightfully typed with weak types.
+
+To circumvent this difficulty, the type checker consider that all values
+returned by a function might rely on persistent mutable state behind the
+scene and should be given a weak type. This restriction on the type of mutable
+values and function application is called the value restriction. Note that
+this value restriction is conservative: there are situations where the value
+restriction is too cautious and gives a weak type to a value that could be
+safely generalized to a polymorphic type:
+
+\begin{caml_example}
+let not_id = (fun x -> x) (fun x -> x);;
+\end{caml_example}
+
+Quite often, this happens when defining function using higher order function.
+To avoid this problem, a solution is to add an explicit argument to the
+function:
+\begin{caml_example}
+let id_again = fun x -> (fun x -> x) (fun x -> x) x;;
+\end{caml_example}
+With this argument, "id_again" is seen as a function definition by the type
+checker and can therefore be generalized.
+
+This kind of manipulation is called eta-expansion in lambda calculus and
+is sometimes referred under this name.
+
+\subsection{The relaxed value restriction}
+
+There is another partial solution to the problem of unnecessary weak type
+in the type of value, which is implemented directly within the type
+checker. Briefly, it is possible to prove that weak types that only appear
+as type parameters in a covariant position can be safely generalized to
```

Comment:
People sometimes use "positive position" instead of "covariant position", and it makes as much sense when subtyping is not considered. It would be helpful to refer to this other wording at some point.